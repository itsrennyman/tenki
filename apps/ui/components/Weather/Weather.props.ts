export interface WeatherProps {
  isLoading: boolean;
  isError: boolean;
  onPlaceChange?: () => void;
  data: WeatherData | null;
}

export interface WeatherData {
  day: string;
  city: string;
  country: string;
  temperature: string;
  description: string;
  infos: WeatherInfo[];
}

export interface WeatherInfo {
  label: string;
  value: string | number;
}
