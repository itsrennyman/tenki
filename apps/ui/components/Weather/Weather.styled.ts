import styled from 'styled-components';

const StyledWeather = styled.div`
  border-radius: 25px;
  box-shadow: 0 0 70px -10px rgb(0 0 0 / 20%);
  background-color: #222831;
  color: #ffffff;
  height: 400px;
  display: flex;
  flex: 1;
  justify-content: space-between;
  max-width: 600px;
  padding: 0;
`;

export { StyledWeather };
