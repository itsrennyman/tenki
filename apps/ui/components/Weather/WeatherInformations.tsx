import styled from 'styled-components';

const StyledWeatherInformations = styled.div`
  display: flex;
  flex: 1.1;
  flex-direction: column;
  gap: 2rem;
  overflow-y: scroll;
  padding: 35px;
`;

export { StyledWeatherInformations as WeatherInformations };
