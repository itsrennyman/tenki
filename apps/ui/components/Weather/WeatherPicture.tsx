import styled from 'styled-components';

const StyledWeatherPicture = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: space-between;
  align-items: flex-start;
  background-color: black;
  border-radius: 25px;
  padding: 25px;
  transform: scale(1.03);
  transition: all 0.5s cubic-bezier(0, 0.27, 0.17, 1.04);
  background: linear-gradient(
      to bottom,
      rgb(141 141 141 / 52%),
      rgba(117, 19, 93, 0.73)
    ),
    url(https://images.unsplash.com/photo-1559963110-71b394e7494d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=675&q=80);

  &:hover {
    transform: scale(1.05) perspective(1500px) rotateY(10deg);
  }
`;

export { StyledWeatherPicture as WeatherPicture };
