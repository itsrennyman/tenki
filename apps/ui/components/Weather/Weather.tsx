import styled from 'styled-components';
import { Flex } from '../Flex';
import { Ripple } from '../Ripple';
import { Text } from '../Text';
import { WeatherInfo, WeatherProps } from './Weather.props';
import { StyledWeather } from './Weather.styled';

const Button = styled.button`
  background-color: rgb(0 0 0 / 20%);
  border: none;
  padding: 1rem;
  border-radius: 0.5rem;
  font-size: 0.85rem;
  font-weight: 700;
  text-transform: uppercase;
  cursor: pointer;
  outline: none;
  color: #fff;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;

  &:hover:enabled {
    background-color: rgb(0 0 0 / 30%);
    transform: scale(1.02);
  }

  &:disabled {
    background-color: rgb(0 0 0 / 10%);
    cursor: not-allowed;
  }
`;

const Informations = styled.div`
  display: flex;
  flex: 1.1;
  flex-direction: column;
  gap: 2rem;
  overflow-y: scroll;
  padding: 35px;
`;

const Picture = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: space-between;
  align-items: flex-start;
  background-color: black;
  border-radius: 25px;
  padding: 25px;
  transform: scale(1.03);
  transition: all 0.5s cubic-bezier(0, 0.27, 0.17, 1.04);
  background: linear-gradient(
      to bottom,
      rgb(141 141 141 / 52%),
      rgba(117, 19, 93, 0.73)
    ),
    url(https://images.unsplash.com/photo-1559963110-71b394e7494d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=675&q=80);

  &:hover {
    transform: scale(1.05) perspective(1500px) rotateY(10deg);
  }
`;

const Information = ({ label, value }: WeatherInfo) => {
  return (
    <Flex justifyContent="space-between">
      <Text as="dt" textTransform="uppercase" fontWeight={700}>
        {label}
      </Text>
      <Text as="dd">{value}</Text>
    </Flex>
  );
};

const Weather = ({
  data,
  isLoading = false,
  isError = false,
  onReload,
}: WeatherProps) => {
  if (isLoading) {
    return (
      <StyledWeather data-cy="loading">
        <Flex justifyContent="center" alignItems="center" grow={1}>
          <Ripple />
        </Flex>
      </StyledWeather>
    );
  }

  if (isError) {
    return (
      <StyledWeather data-cy="error">
        <Flex justifyContent="center" alignItems="center" grow={1}>
          Impossible to load weather data.
        </Flex>
      </StyledWeather>
    );
  }

  const { infos, ...rest } = data;

  const informations = infos.map((element, index) => {
    return <Information key={index} {...element} />;
  });

  return (
    <StyledWeather data-testid="weather" data-cy="weather">
      <Picture>
        <Text>{rest.day}</Text>
        <div>
          <Text>{rest.city}</Text>
          <Text fontSize="4em" fontWeight={700}>
            {rest.temperature}
          </Text>
          <Text fontSize="1em" fontWeight={600}>
            {rest.description}
          </Text>
        </div>
      </Picture>

      <Informations>
        <Flex direction="column" gap="1rem" grow={1}>
          {informations}
        </Flex>

        <Button onClick={onReload}>Reload Data!</Button>
      </Informations>
    </StyledWeather>
  );
};

export { Weather };
