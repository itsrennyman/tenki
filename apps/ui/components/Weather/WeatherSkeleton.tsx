import { Button } from '../Button';
import { Flex } from '../Flex';
import { Skeleton } from '../Skeleton';
import { Text } from '../Text';
import { WeatherInfo } from './Weather.props';
import { StyledWeather } from './Weather.styled';
import { WeatherInformations } from './WeatherInformations';
import { WeatherPicture } from './WeatherPicture';

const data = {
  day: 'Monday',
  city: 'Milan',
  country: 'Italy',
  temperature: 25,
  description: 'Sunny',
  infos: [
    { label: 'Wind Speed', value: '10 km/h' },
    { label: 'Wind Direction', value: 'SSE' },
    { label: 'UV Index', value: '10' },
    { label: 'Humidity', value: '90%' },
    { label: 'Pressure', value: '1021mb' },
    { label: 'Visibility', value: '16km' },
  ],
};

const InformationSkeleton = ({ label, value }: WeatherInfo) => {
  return (
    <Flex justifyContent="space-between">
      <Skeleton>
        <Text as="dt" textTransform="uppercase" fontWeight={700}>
          {label}
        </Text>
      </Skeleton>
      <Skeleton>
        <Text as="dd">{value}</Text>
      </Skeleton>
    </Flex>
  );
};

const WeatherSkeleton = () => {
  const { infos } = data;

  const informations = infos.map((element, index) => {
    return <InformationSkeleton key={index} {...element} />;
  });

  return (
    <StyledWeather data-cy="loading">
      <WeatherPicture />

      <WeatherInformations>
        <Flex direction="column" gap="1rem" grow={1}>
          {informations}
        </Flex>

        <Skeleton>
          <Button disabled={true}>Another Place!</Button>
        </Skeleton>
      </WeatherInformations>
    </StyledWeather>
  );
};

export { WeatherSkeleton };
