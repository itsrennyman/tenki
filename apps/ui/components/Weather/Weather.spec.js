import '@testing-library/dom';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TestRenderer from 'react-test-renderer';
import { Weather } from './Weather';

const props = {
  day: 'Monday',
  city: 'Milan',
  country: 'Italy',
  temperature: 25,
  description: 'Sunny',
  infos: [
    { label: 'Wind', value: '10km/h' },
    { label: 'Humidity', value: '60%' },
    { label: 'Pressure', value: '1025mb' },
    { label: 'Sunrise', value: '6:00am' },
    { label: 'Sunset', value: '6:00pm' },
    { label: 'Geo coords', value: '45.465,12.234' },
  ],
};

test('should render without crashing', () => {
  render(<Weather data={props} />);
  expect(screen.getByTestId('weather')).toBeInTheDocument();
});

test('should render correctly', () => {
  const tree = TestRenderer.create(<Weather data={props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('should trigger correctly onPlaceChange on button click', async () => {
  const user = userEvent.setup();
  const onPlaceChange = jest.fn();
  render(<Weather data={props} onPlaceChange={onPlaceChange} />);
  await user.click(screen.getByRole('button'));
  expect(onPlaceChange).toHaveBeenCalled();
});
