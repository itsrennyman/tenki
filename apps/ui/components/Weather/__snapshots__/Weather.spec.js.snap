// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="sc-hKMtZM eiqwmw sc-eCYdqJ ktzijR"
  data-cy="weather"
  data-testid="weather"
>
  <div
    className="sc-iBkjds iJWbDD"
  >
    <div
      className="sc-bczRLJ bDDrHP"
      direction="column"
    >
      <p
        className="sc-dkzDqf cqLQeI"
      >
        Monday
      </p>
    </div>
    <div>
      <p
        className="sc-dkzDqf fhmgnY"
        fontSize="0.9em"
      >
        Milan
         (
        Italy
        )
      </p>
      <p
        className="sc-dkzDqf brehns"
        fontSize="4em"
        fontWeight={700}
      >
        25
      </p>
      <p
        className="sc-dkzDqf fgcoac"
        fontSize="1em"
        fontWeight={600}
      >
        Sunny
      </p>
    </div>
  </div>
  <div
    className="sc-gKXOVf OOeIB"
  >
    <div
      className="sc-bczRLJ dWSqhE"
      direction="column"
    >
      <div
        className="sc-bczRLJ kwAZZz"
      >
        <dt
          className="sc-dkzDqf gHxGtr"
          fontWeight={700}
        >
          Wind
        </dt>
        <dd
          className="sc-dkzDqf cqLQeI"
        >
          10km/h
        </dd>
      </div>
      <div
        className="sc-bczRLJ kwAZZz"
      >
        <dt
          className="sc-dkzDqf gHxGtr"
          fontWeight={700}
        >
          Humidity
        </dt>
        <dd
          className="sc-dkzDqf cqLQeI"
        >
          60%
        </dd>
      </div>
      <div
        className="sc-bczRLJ kwAZZz"
      >
        <dt
          className="sc-dkzDqf gHxGtr"
          fontWeight={700}
        >
          Pressure
        </dt>
        <dd
          className="sc-dkzDqf cqLQeI"
        >
          1025mb
        </dd>
      </div>
      <div
        className="sc-bczRLJ kwAZZz"
      >
        <dt
          className="sc-dkzDqf gHxGtr"
          fontWeight={700}
        >
          Sunrise
        </dt>
        <dd
          className="sc-dkzDqf cqLQeI"
        >
          6:00am
        </dd>
      </div>
      <div
        className="sc-bczRLJ kwAZZz"
      >
        <dt
          className="sc-dkzDqf gHxGtr"
          fontWeight={700}
        >
          Sunset
        </dt>
        <dd
          className="sc-dkzDqf cqLQeI"
        >
          6:00pm
        </dd>
      </div>
      <div
        className="sc-bczRLJ kwAZZz"
      >
        <dt
          className="sc-dkzDqf gHxGtr"
          fontWeight={700}
        >
          Geo coords
        </dt>
        <dd
          className="sc-dkzDqf cqLQeI"
        >
          45.465,12.234
        </dd>
      </div>
    </div>
    <button
      className="sc-jSMfEi PCDIE"
    >
      Another Place!
    </button>
  </div>
</div>
`;
